README.md

```diff
@@ -1,3 +1,270 @@
+
+``` go
+	// NewWatchSource also has watch options that can be passed to have the
+	// ability to use a ticker for polling, set a logger, and more
+	watchingFileSource, fsErr := file.NewWatchingSource(
+		"path/to/config", &yaml.Decoder{})
+
+	if fsErr != nil {
+		// error handling
+		return
+	}
+
+	// additional sources can be passed along with the watching file source and the
+	// precedence order will still be dictated by the order in which the sources are
+	// defined in the Config function.
+	d, err := dials.Config(context.Background(), config, watchingFileSource)
+	if err != nil {
+		// error handling
+	}
+
+	d.Fill(config)
+```
+
+### Source
+The Source interface is implemented by different configuration sources that populate the configuration struct. Dials currently supports environment variables, command line flags, and config file sources. When `dials.Config` function is going through the different sources to extract the values, it calls the `Value` method on each of these sources. This allows for the logic of the Source to be encapsulated while giving the application access to the values populated by each Source. Please note that the Value method on the Source interface and the Watcher interface are likely to change in the near future.
+
+
+### Decoder
+Decoders are modular allowing users to mix and match Decoders and Sources. Dials currently supports Decoders that decode different data formats (JSON, YAML, and TOML) and insert the values into the appropriate fields in the config struct. Decoders can be expanded from that use case and users can write their own Decoders to perform the tasks they like (more info in the section below). Decoder is called when the supported Source calls the `Decode` method to unmarshal the data into the config struct and returns the populated struct. There are two sources that the Decoders can be used with: files (including watched files) and `static.StringSource`. Please note that the Decoder interface is likely to change in the near future.
+
+### Write your own Source and Decoder
+If you wish to define your own source, implement the `Source` interface and pass the source to the `dials.Config` function. If you want the Source to interact with a Decoder, call `Decode` in the `Value` method of the Source. If you want to define your own Decoder to interact with your Source, implement the `Decoder` interface. Since Decoders are modular, keep the logic of Decoder encapsulated and separate from the Source. For example, you can have an `HTTP` Source and custom `Get` and `Post` Decoders that repeatedly make `GET` and `POST` requests. The user can use the HTTP Source with either one of the Get or Post decoders and potentially also use the `POST` Decoder with other Sources.
```

Comment:
This is confusing.
The `Decoder` wouldn't make `GET` or `POST` requests. The `Decoder` would process the response that the `Source` got from one of those requests.

Generally, defining different `Source` and `Decoder` implementations should only be done where their operation is orthogonal, such as the File source, or an HTTP fetching source, where they get back bytes that must then be converted to a struct by *something*.

I'd avoid discussing "writing a `Decoder` for your `Source`" since that implies non-orthogonality between the implementations, in which case the decoding should probably be part of the `Source` itself. (or can be another implementation-specific struct/parameter)